### ISSUE TITLE
build cpu with MKL Lapack from source code

### ISSUE THREAD

**Issue by AgapovO** [arrayfire/arrayfire#949]
An error appears during the assembly of the project from source code when using MKL as lapack library. Declaration of MKL lapack functions (mkl_lapack.h) do not coincide with the functions which are used in the arrayfire lib.

**Comment by pavanky** [comment]
@AgapovO Our CMake file does not automatically find MKL. For now you have to set things up manually. We need to resolve this at some point. 

For reference, this is what we use for our binary installers:

On Linux:

```
-DBUILD_CPU=ON -DBUILD_CUDA=ON -DCOMPUTES_DETECTED_LIST="20;30;35;50;52" -DBUILD_OPENCL=ON -DBUILD_GRAPHICS=ON -DUSE_GLEWmx_STATIC=ON -DBUILD_EXAMPLES=ON -DBUILD_TEST=ON -DBUILD_DOCS=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo  -DUSE_CPU_MKL=ON -DCBLAS_LIBRARIES="/opt/intel/mkl/lib/intel64/libmkl_core.a;/opt/intel/mkl/lib/intel64/libmkl_gnu_thread.a;/opt/intel/mkl/lib/intel64/libmkl_intel_lp64.a" -DFFTW_LIB="/opt/intel/mkl/lib/intel64/libmkl_core.a;/opt/intel/mkl/lib/intel64/libmkl_gnu_thread.a;/opt/intel/mkl/lib/intel64/libmkl_intel_lp64.a" -DFFTWF_LIB="/opt/intel/mkl/lib/intel64/libmkl_core.a;/opt/intel/mkl/lib/intel64/libmkl_gnu_thread.a;/opt/intel/mkl/lib/intel64/libmkl_intel_lp64.a" -DLAPACKE_LIB="/opt/intel/mkl/lib/intel64/libmkl_core.a;/opt/intel/mkl/lib/intel64/libmkl_gnu_thread.a;/opt/intel/mkl/lib/intel64/libmkl_intel_lp64.a" -DLAPACK_LIB="/opt/intel/mkl/lib/intel64/libmkl_core.a;/opt/intel/mkl/lib/intel64/libmkl_gnu_thread.a;/opt/intel/mkl/lib/intel64/libmkl_intel_lp64.a;/opt/intel/mkl/lib/intel64/libmkl_intel_lp64.a;/usr/lib64/libgomp.so.1;/usr/lib64/libdl.so;/usr/lib64/libpthread.so;/usr/lib/libm.so" -DFFTW_INCLUDES=/opt/intel/mkl/include/fftw/ -DLAPACKE_INCLUDES=/opt/intel/mkl/include/
```

On Windows:

```
-G "Visual Studio 12 2013 Win64" -T "Intel C++ Compiler XE 15.0" -DCMAKE_CXX_COMPILER:STRING="C:/Program Files (x86)/Intel/Composer XE 2015/bin/intel64/icl.exe" -DCMAKE_C_COMPILER:STRING="C:/Program Files (x86)/Intel/Composer XE 2015/bin/intel64/icl.exe" -DCMAKE_BUILD_TYPE:STRING="RelWithDebInfo" -DCMAKE_INSTALL_PREFIX:STRING="package" ^
-DBUILD_EXAMPLES:STRING=ON -DBUILD_TEST:STRING=OFF -DBUILD_DOCS:STRING=ON ^
-DFREEIMAGE_FOUND:STRING=ON -DUSE_FREEIMAGE_STATIC:BOOL=ON -DFREEIMAGE_INCLUDE_PATH:STRING="%FI_DIR%" -DFREEIMAGE_STATIC_LIBRARY:STRING="%FI_DIR%/FreeImageLib.lib" -DFREEIMAGE_DYNAMIC_LIBRARY:STRING="%FI_DIR%/FreeImage.lib" ^
-DBUILD_CPU:STRING=ON -DFFTW_ROOT:STRING="%MKL_DIR%" -DFFTW_LIBRARIES:STRING="%MKL_DIR%/lib/mkl_rt.lib" -DFFTW_LIB:STRING="" -DFFTWF_LIB:STRING="" ^
-DUSE_CPU_MKL:BOOL=ON -DCBLAS_INCLUDE_DIR:STRING="%MKL_DIR%/include" -DCBLAS_cblas_LIBRARY:STRING="%MKL_DIR%/lib/mkl_rt.lib" ^
-DLAPACKE_INCLUDES:STRING="%MKL_DIR%/include" -DLAPACKE_LIB:STRING="%MKL_DIR%/lib/mkl_rt.lib;" -DLAPACK_LIB:STRING="%MKL_DIR%/lib/mkl_rt.lib" ^
-DBUILD_CUDA:STRING=ON -DCUDA_TOOLKIT_ROOT_DIR:STRING="%CUDA_PATH_V7_0%" -DCUDA_COMPUTE_CAPABILITY:STRING=20 ^
-DBUILD_OPENCL:STRING=OFF -DUSE_OPENCL_MKL:BOOL=ON -DBOOST_INCLUDEDIR:STRING="%DEPS_DIR%/boost_1_56_0"
```

**Comment by AgapovO** [comment]
"Our CMake file does not automatically find MKL" 

I'm not talking about that. I linked MKL libes. But when I builded the project error occurs.  Declaration of MKL lapack functions (mkl_lapack.h) do not coincide with the functions which are used in the arrayfire lib.   For example. A prototype is used in the file "cholesky.cpp" is 

    int LAPACKE_dpotrf(ORDER_TYPE, char, int, T*, int);. 

But same prototype in the file "mkl_lapacke.h"  is

    void dpotrf( const char* uplo, const MKL_INT* n, double* a, const MKL_INT* lda,  MKL_INT* info );

Prototypes are not the same and error occurs.

**Comment by pavanky** [comment]
You need to pass "-DUSE_CPU_MKL=ON" flag to cmake for CPU backend and "-DUSE_OPENCL_MKL=ON" to cmake for compiling OpenCL backend.

**Comment by AgapovO** [comment]
I build on windows using scripts. I attached my "common.bat". 

```Batch
REM DIRECTORIES
REM USE DOUBLE QUOTES IF PATHS INCLUDE SPACES
SET WORKSPACE=D:/workspace
SET AF_DIR=%WORKSPACE%/arrayfire
REM AF_INSTALL_PATH is where ArrayFire files are installed.
REM AF_INSTALL_PATH can be relative path from AF_DIR/build or absolute path
SET AF_INSTALL_PATH="package"
SET DEPS_DIR=%WORKSPACE%/dependencies
REM Relative to AF_DIR
SET BUILD_DIR=build
REM BUILD_TYPE can be Release, Debug, RelWithDebWinfo, MinSizeRel
SET BUILD_TYPE=Release

REM ARRAYFIRE BUILD OPTIONS
REM SELECT WHICH BACKENDS TO BUILD BY SETTING THEM TO ON
SET CPU=ON
SET CUDA=OFF
SET OPENCL=OFF
SET GRAPHICS=OFF

REM Set CUDA_COMPUTE_DETECT to OFF to disable auto detection.
REM If no CUDA computes are detected (for example when running on a remote machine)
REM it will default to 20, 30, 50, unless at least one of CUDA_COMPUTE_XY (below) are set to ON
SET CUDA_COMPUTE_DETECT=ON

REM Set these to ON for adding a architecture while compiling.
REM If CUDA_COMPUTE_DETECT is turned on, the detected computes will be enabled automatically and override any OFF computes.
SET CUDA_COMPUTE_20=OFF
SET CUDA_COMPUTE_30=OFF
SET CUDA_COMPUTE_35=OFF
SET CUDA_COMPUTE_50=OFF
SET CUDA_COMPUTE_52=OFF

REM TOGGLE EXAMPLES AND TESTS
SET EXAMPLES=OFF
SET TESTS=OFF

REM FREEIMAGE_TYPE can be OFF, STATIC, DYNAMIC
SET FREEIMAGE_TYPE=STATIC

REM CPU_FFT_TYPE can be FFTW, ACML, MKL
REM Note: ACML 6.1 has known failure issues
SET CPU_FFT_TYPE=MKL

REM CPU_BLAS_TYPE can be LAPACKE, MKL
REM Ideally, BLAS_TYPE and LAPACK_TYPE should be the same
SET CPU_BLAS_TYPE=MKL

REM CPU_LAPACK_TYPE can be LAPACKE, MKL
SET CPU_LAPACK_TYPE=MKL

REM GLEWmx_STATIC can be ON or OFF
SET GLEWmx_STATIC=OFF

REM PATH SETTING TO RUN EXECUTABLES
SET FI_DIR=%DEPS_DIR%/freeimage-3.17.0_x64
SET FFTW_DIR=%DEPS_DIR%/fftw-3.3.4
SET ACML_DIR=%DEPS_DIR%/acml/ifort64
SET MKL_DIR=%DEPS_DIR%/mkl
SET LAPACKE_DIR=%DEPS_DIR%/lapacke
SET GLEW_DIR=%DEPS_DIR%/glew
SET GLFW_DIR=%DEPS_DIR%/glfw

SET PATH_EXT=%AF_DIR%\%BUILD_DIR%\package\lib;

REM FREEIMAGE
SET FREEIMAGE_OPTIONS=
if "%FREEIMAGE_TYPE%"=="DYNAMIC" (
    SET FREEIMAGE_OPTIONS=-DFREEIMAGE_FOUND:STRING=ON -DUSE_FREEIMAGE_STATIC:BOOL=OFF -DFREEIMAGE_INCLUDE_PATH:STRING="%FI_DIR%" -DFREEIMAGE_STATIC_LIBRARY:STRING="%FI_DIR%/FreeImageLib.lib" -DFREEIMAGE_DYNAMIC_LIBRARY:STRING="%FI_DIR%/FreeImage.lib"
    SET PATH_EXT=%PATH_EXT%%FI_DIR%;
) else ( REM STATIC
    SET FREEIMAGE_OPTIONS=-DFREEIMAGE_FOUND:STRING=ON -DUSE_FREEIMAGE_STATIC:BOOL=ON -DFREEIMAGE_INCLUDE_PATH:STRING="%FI_DIR%" -DFREEIMAGE_STATIC_LIBRARY:STRING="%FI_DIR%/FreeImageLib.lib" -DFREEIMAGE_DYNAMIC_LIBRARY:STRING="%FI_DIR%/FreeImage.lib"
)

REM FFT
SET FFT_OPTIONS=
if "%CPU_FFT_TYPE%"=="FFTW" (
    SET FFT_OPTIONS=-DFFTW_ROOT:STRING="%FFTW_DIR%" -DFFTW_LIB:STRING="%FFTW_DIR%/libfftw3-3.lib" -DFFTWF_LIB:STRING="%FFTW_DIR%/libfftw3f-3.lib" -DFFTWL_LIB:STRING="%FFTW_DIR%/libfftw3l-3.lib"
    SET PATH_EXT=%PATH_EXT%%FFTW_DIR%;
) else (
    if "%CPU_FFT_TYPE%"=="ACML" (
        REM SET ACML PATH HERE
        SET FFT_OPTIONS=-DFFTW_ROOT:STRING="%ACML_DIR%" -DFFTW_LIBRARIES:STRING="%ACML_DIR%/lib/acml_fftw.lib" -DFFTW_LIB:STRING="" -DFFTWF_LIB:STRING=""
        SET PATH_EXT=%PATH_EXT%%ACML_DIR%\lib;
    ) else (
        REM SET MKL PATH HERE
        SET FFT_OPTIONS=-DFFTW_ROOT:STRING="%MKL_DIR%" -DFFTW_LIBRARIES:STRING="%MKL_DIR%/lib/mkl_rt.lib" -DFFTW_LIB:STRING="" -DFFTWF_LIB:STRING=""
        SET PATH_EXT=%PATH_EXT%%MKL_DIR%\bin;
    )
)

SET BLAS_OPTIONS=
if "%CPU_BLAS_TYPE%"=="LAPACKE" (
    REM SET LAPACKE PATH HERE
    SET BLAS_OPTIONS=-DUSE_CPU_F77_BLAS:BOOL=ON -DCBLAS_INCLUDE_DIR:STRING="%LAPACKE_DIR%/include" -DCBLAS_cblas_LIBRARY:STRING="%LAPACKE_DIR%/lib/libblas.lib"
    SET PATH_EXT=%PATH_EXT%%LAPACKE_DIR%\bin;
) else (
    REM SET MKL PATH HERE
    SET BLAS_OPTIONS=-DUSE_CPU_MKL:BOOL=ON -DUSE_OPENCL_MKL:BOOL=ON -DCBLAS_INCLUDE_DIR:STRING="%MKL_DIR%/include" -DCBLAS_cblas_LIBRARY:STRING="%MKL_DIR%/lib/mkl_rt.lib"
    SET PATH_EXT=%PATH_EXT%%MKL_DIR%\bin;
)

SET LAPACK_OPTIONS=
if "%CPU_LAPACK_TYPE%"=="LAPACKE" (
    REM SET LAPACKE PATH HERE
    SET LAPACK_OPTIONS=-DLAPACKE_ROOT:STRING="%LAPACKE_DIR%"
    REM SET LAPACK_OPTIONS=-DLAPACK_INCLUDE_DIR:STRING="%LAPACKE_DIR%/include" -DLAPACKE_LIB:STRING="%LAPACKE_DIR%/lib/liblapacke.lib" -DLAPACK_LIB:STRING="%LAPACKE_DIR%/lib/liblapack.lib"
    SET PATH_EXT=%PATH_EXT%%LAPACKE_DIR%\bin;
) else (
    REM SET MKL PATH HERE
    SET LAPACK_OPTIONS=-DUSE_CPU_MKL:BOOL=ON -DUSE_OPENCL_MKL:BOOL=ON -DLAPACKE_INCLUDES:STRING="%MKL_DIR%/include" -DLAPACKE_LIB:STRING="%MKL_DIR%/lib/mkl_rt.lib;" -DLAPACK_LIB:STRING="%MKL_DIR%/lib/mkl_rt.lib"
    SET PATH_EXT=%PATH_EXT%%MKL_DIR%\bin;
)

SET GRAPHICS_OPTIONS=
if "%GRAPHICS%"=="ON" (
    SET GRAPHICS_OPTIONS=-DBUILD_GRAPHICS:BOOL=ON -DGLEW_ROOT_DIR="%GLEW_DIR%" -DUSE_GLEWmx_STATIC:BOOL=%GLEWmx_STATIC% -DGLEW_INCLUDE_DIR:STRING="%GLEW_DIR%/include" -DGLEW_LIBRARY="%GLEW_DIR%/lib/glew32.lib" -DGLEWmxd_LIBRARY="%GLEW_DIR%/lib/glew32mx.lib" -DGLEWmxs_LIBRARY="%GLEW_DIR%/lib/glew32mxs.lib" -DGLFW_ROOT_DIR="%GLFW_DIR%" -DGLFW_INCLUDE_DIR:STRING="%GLFW_DIR%/include" -DGLFW_LIBRARY="%GLFW_DIR%/lib/glfw3.lib"
    SET PATH_EXT=%PATH_EXT%%DEPS_DIR%\glew\lib;%DEPS_DIR%\glfw\lib;
) else (
    SET GRAPHICS_OPTIONS=-DBUILD_GRAPHICS:BOOL=OFF
)

if "%CUDA%"=="ON" (
    SET PATH_EXT=%PATH_EXT%%CUDA_PATH%\bin;%CUDA_PATH%\nvvm\bin;
)

REM TOOLS CONFIGURATION
SET THREADS=8
SET MSBUILD="C:\Program Files (x86)\MSBuild\12.0\Bin\MSBuild.exe" /m:%THREADS%
SET GIT_EXE="git.exe"
SET CMAKE="D:\Program Files (x86)\CMake\bin\cmake"
SET CMAKE_GENERATOR=-G "Visual Studio 12 2013 Win64"
SET CTEST="C:\Program Files (x86)\CMake\bin\ctest.exe" --force-new-ctest-process --output-on-failure
```

**Comment by pavanky** [comment]
@shehzan10 can help you out with windows.

**Comment by AgapovO** [comment]
Thank you. I will ask him about it.

**Comment by pavanky** [comment]
He will reply here soon. I was just getting his attention to this issue :)

**Comment by shehzan10** [comment]
So we use lapacke rather than lapack. The only place where we use lapack is OSX, which for now is not relevant to this discussion.

I'm curious as to why you found the need for mkl_lapack.h while building. Our source code only includes mkl_lapacke.h. Can you tell me more about this error?

**Comment by AgapovO** [comment]
I use MKL 10.2.6.037. I can't find "mkl_lapacke.h" and manually change "mkl_lapacke.h" to "mkl_lapack.h" in your source code. I think that's why there was the error. Tell me where to find mkl_lapacke.h.

**Comment by shehzan10** [comment]
That makes more sense.

We use the Intel Composer 2015 at our end which includes MKL. I'm not sure how the versions correspond to yours. A quick search led me to this (https://software.intel.com/en-us/articles/which-version-of-the-intel-ipp-intel-mkl-and-intel-tbb-libraries-are-included-in-the-intel) page which says composer xe has version 11. I tried looking up where lapacke was added, and it seems like it was 10.3 update 5 (https://software.intel.com/en-us/articles/intel-mkl-103-release-notes).

lapacke.h is located at `C:\Program Files (x86)\Intel\Composer XE 2015\mkl\include\mkl_lapacke.h` on our systems.

Even if we do resolve the lapacke.h, it is crucial for you to have the mkl libraries that support lapacke.

Is there a reason why you are not using the installer?
